File: interview_prep/concurrency/thread_creation.cpp
// Creating and using std::thread.
#include <iostream>
#include <thread>
#include <vector>

int main() {
    std::thread worker([]{ std::cout << "worker thread id=" << std::this_thread::get_id() << '\n'; });
    worker.join();

    std::vector<std::thread> pool;
    for (int i = 0; i < 4; ++i) pool.emplace_back([i]{ std::cout << "task " << i << '\n'; });
    for (auto &t : pool) t.join();
    return 0;
}

File: interview_prep/deep_dives/26_os_threads/thread_creation.py
"""Measure thread creation cost vs a thread pool."""
import threading
import time
from concurrent.futures import ThreadPoolExecutor


def task() -> int:
    return sum(range(1000))


def with_threads(n: int) -> tuple[int, float]:
    start = time.perf_counter()
    results = []
    threads = [threading.Thread(target=lambda: results.append(task())) for _ in range(n)]
    for t in threads:
        t.start()
    for t in threads:
        t.join()
    return len(results), time.perf_counter() - start


def with_pool(n: int) -> tuple[int, float]:
    start = time.perf_counter()
    with ThreadPoolExecutor(max_workers=8) as pool:
        results = list(pool.map(lambda _: task(), range(n)))
    return len(results), time.perf_counter() - start


if __name__ == "__main__":
    assert with_threads(50)[0] == 50
    assert with_pool(50)[0] == 50
    print("thread creation ok")
